assimilation_code/programs/preprocess/preprocess.f90

```diff
@@ -519,210 +577,716 @@ program preprocess
 ! file and then proceed.
 
 ! There are seven special code sections (ITEMS) in the obs_def file at present.
-! Each copies code in from the special type specific obs_kind modules
-! Loop goes to N+1 so that lines after the last item are copied to the output.
-ITEMS: do i = 1, 8
-   READ_LINE: do
-      read(obs_def_in_unit, 222, IOSTAT = ierr) line
-      222 format(A256)
-
-      ! Check for end of file (it's an error if this is before the 
-      ! 7 DART ITEMS have been passed)
-      if(ierr /= 0) then
-         if(i < 8) then
-            call error_handler(E_ERR, 'preprocess', &
-               'Input DEFAULT obs_def file ended unexpectedly', &
-               source, revision, revdate)
-         else
-            exit ITEMS
-         endif
-      endif
-
-      ! Check to see if this line indicates the start of an insertion section
-      test = adjustl(line)
-      t_string = '! DART PREPROCESS ' // trim(preprocess_string(i)) // ' INSERTED HERE'
-      if(trim(test) == trim(t_string)) exit READ_LINE
-
-      ! Write this line into the output file
-      write(obs_def_out_unit, 21) trim(line)
-      21 format(A)
+! Each copies code in from the special type specific obs_qty modules
+linenum4 = 0
+ITEMS: do i = 1, NUM_SECTIONS
 
-   end do READ_LINE
+   t_string = '! DART PREPROCESS ' // trim(preprocess_string(i)) // ' INSERTED HERE'
+   call copy_until(obs_def_in_unit, input_obs_def_mod_file, t_string, linenum4, &
+                   obs_def_out_unit, output_obs_def_mod_file, .false.)
 
    ! The 'USE FOR OBS_QTY_MOD' section is handled differently; lines are not
```

Comment:
this is dependent on QTY not KIND.  may need to be changed depending on decisions about string contents.

edit to add - perhaps not.  QTY_ITEM is a parameter and doesn't necessarily have to match the choices for the text line.